iquip/apps/dataviewer.py

```diff
@@ -820,69 +810,127 @@ class _DatasetFetcherThread(QThread):
+        
+        Args:
+            path: The API path, i.e., the request url is "http://{ip}:{port}/{path}".
+            params: Params argument for the GET request.
+            default: The return value of this method when an exception occurs
+              during the GET request.
+            timeout: Timeout argument for the GET request.
+        """
+        url = f"http://{self.ip}:{self.port}/{path}"
         try:
-            response = requests.get(f"http://{self.ip}:{self.port}/dataset/master/",
-                                    params={"key": f"{self.name}.units"},
-                                    timeout=10)
+            response = requests.get(url, params=params, timeout=timeout)
             response.raise_for_status()
         except requests.exceptions.RequestException:
-            logger.exception("Failed to fetch the dataset units.")
-            units = [None] * (dataset.shape[1] - 1)
+            logger.exception("Failed to GET %s", url)
+            return default
+        return response.json()
+
+    def _initialize(self) -> float:
+        """Fetches the target dataset to initialize the local dataset.
+        
+        Returns:
+            The received timestamp or -1 if it failed.
+        """
+        response = self._get("dataset/master/", {"key": self.name})
+        if response is None or response[0] < 0:
+            return -1
+        timestamp, rawDataset = response
+        dataset = np.array(rawDataset)
+        numberOfParameters = dataset.shape[1] if dataset.ndim > 1 else 0
+        _, parameters = self._get(
+            "dataset/master/",
+            {"key": f"{self.name}.parameters"},
+            (0, list(map(str, range(numberOfParameters)))),
+        )
+        rawUnits = self._get("dataset/master/", {"key": f"{self.name}.units"})
+        if rawUnits is None:
+            units = [None] * (numberOfParameters)
         else:
-            units = [unit if unit else None for unit in response.json()]
-        self.fetched.emit(dataset, parameters, units)
+            units = [unit if unit else None for unit in rawUnits[1]]
+        self.initialized.emit(dataset, parameters, units)
+        return timestamp
+
+    def stop(self):
+        """Stops the thread."""
+        self._running = False
+
+    def run(self):
+        """Overridden."""
+        timestamp = self._initialize()
+        if timestamp < 0:
+            self.stopped.emit("Failed to get dataset.")
+            return
+        url = "dataset/master/modification"
+        params = {"key": self.name, "timestamp": timestamp, "timeout": 10}
+        while self._running:
+            response = self._get(url, params, timeout=12)
+            if response is None:
+                self.stopped.emit("Failed to get modifications.")
+                return
+            timestamp, modifications = response
+            if timestamp < 0:
+                timestamp = self._initialize()
+                if timestamp < 0:
```

Comment:
Is it right that I understood?

First, if the dataset was removed, try to get it again.
Then, if there is the new dataset, run without error.
Otherwise, i.e. there is no dataset still, emit `stopped` signal.

Reply:
Yes, that's exactly how the `last_deleted` timestamp works.
Datasets are deleted when it is explicitly deleted or overwritten.